### PR TITLE
fix: `distributed.all_gather()` support for uneven tensors by explicit padding

### PR DESCRIPTION
**Pull request by fprill**
## Description

`distributed.all_gather()` support for uneven tensors does not apply to all backends
see
- https://docs.pytorch.org/docs/stable/distributed.html#torch.distributed.all_gather
- https://github.com/pytorch/pytorch/issues/156627

##  Additional notes ##

@japols

### CODE REVIEW

File: models/src/anemoi/models/distributed/primitives.py [hunk 1]
```diff
@@ -105,19 +105,23 @@ def _gather(
...
         tensor_list = [
             torch.empty(
-                shapes[rank], dtype=input_.dtype, layout=input_.layout, device=input_.device, memory_format=input_format
+                max_shape, dtype=input_.dtype, layout=input_.layout, device=input_.device, memory_format=input_format
             )
-            for rank in range(comm_size)
+            for _ in range(comm_size)
         ]
 
+        pad_size = max_shape[1] - input_.shape[1]
```
Comment (ssmmnn11) [inline models/src/anemoi/models/distributed/primitives.py]: Hardcoded for a specific dim?

File: models/src/anemoi/models/distributed/primitives.py [hunk 2]
```diff
@@ -105,19 +105,23 @@ def _gather(
...
+        input_ = torch.nn.functional.pad(input_, (0, pad_size), mode="constant", value=0)
         tensor_list[comm_rank] = input_
+
         if gather_in_backward:
             dist.all_gather(tensor_list, input_, group=group)
 
         # Note: torch.cat already creates a contiguous tensor.
-        output = torch.cat(tensor_list, dim=dim_).contiguous(memory_format=input_format)
```
Comment (ssmmnn11) [inline models/src/anemoi/models/distributed/primitives.py:118]: why drop memory format?

Comment (fprill) [reply models/src/anemoi/models/distributed/primitives.py:118]: Should be resolved by b450834df8a2be06c7af8e3affb643d06d7c609d

File: models/src/anemoi/models/distributed/primitives.py (hunk 1 above)
Comment (fprill) [reply models/src/anemoi/models/distributed/primitives.py]: should be resolved by f5adc7fa29e33268995ee2aca050729c808dbc50

File: models/src/anemoi/models/distributed/primitives.py [hunk 3]
```diff
@@ -76,48 +158,22 @@ def _gather(
     # See the License for the specific language governing permissions and
     # limitations under the License.
 
-    # get input format
-    input_format = get_memory_format(input_)
-
-    comm_size = dist.get_world_size(group=group)
     # Bypass the function if we are using only 1 GPU.
-    if comm_size == 1:
+    if dist.get_world_size == 1:
```
Comment (japols) [inline models/src/anemoi/models/distributed/primitives.py]:
```suggestion
    if dist.get_world_size(group=group) == 1:
```
